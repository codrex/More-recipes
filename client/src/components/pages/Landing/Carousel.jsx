import React from 'react';
import PropTypes from 'prop-types';
import {
  CAROUSEL_PIX_ONE,
  CAROUSEL_PIX_TWO
} from '../../../constants';

/**
 * Carousel component
 * @param {object} props
 * @return {React} react element
 */
const Carousel = props => (
  <div id="carousel" className={`carousel slide ${props.className}`} data-ride="carousel">
    <ol className="carousel-indicators">
      <li data-target="#carousel" data-slide-to="0" className="active" />
      <li data-target="#carousel" data-slide-to="1" />
    </ol>
    <div className="carousel-inner hero">
      <div className="carousel-item backdrop active">
        <img className="d-block img-fluid" src={CAROUSEL_PIX_ONE} alt="First slide" />
      </div>
      <div className="carousel-item backdrop">
        <img className="d-block img-fluid" src={CAROUSEL_PIX_TWO} alt="First slide" />
      </div>
      {props.children}
    </div>
  </div>
);

Carousel.defaultProps = {
  className: ''
};

Carousel.propTypes = {
  children: PropTypes.objectOf(PropTypes.shape).isRequired,
  className: PropTypes.string,
};

export default Carousel;
