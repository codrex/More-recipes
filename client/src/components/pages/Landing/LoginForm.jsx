import React from 'react';
import PropTypes from 'prop-types';
import { Field, reduxForm } from 'redux-form';
import { Form, Input } from '../../common/FormElements';
import { username, password } from '../../../utils/validator';


/**
 * User login form
 * @param {object} props
 * @return {React} react element
 */
class LoginForm extends React.Component {
  /**
   * constructor
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.login = this.login.bind(this);
  }

  /**
   * login
   * @param {object} value (form values)
   * @return {undefined} undefined
  */
  login(value) {
    this.props.login(value);
  }

  /**
   * render
   * @returns {React} react element
   */
  render() {
    const { handleSubmit, submitting } = this.props;
    return (
      <Form
        submitBtnText={submitting ? 'loading...' : 'Login'}
        onSubmit={handleSubmit(this.props.login)}
        disabled={submitting}
        secondary
      >
        <Field
          component={Input}
          name="username"
          type="text"
          id="username"
          placeholder="Enter username"
          validate={username}
        />
        <Field
          component={Input}
          name="password"
          type="password"
          id="password"
          placeholder="Enter password"
          validate={password}
        />
      </Form>
    );
  }
}
LoginForm.propTypes = {
  login: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  submitting: PropTypes.bool.isRequired,
};

export { LoginForm as PureLoginForm };

export default reduxForm({
  form: 'LoginForm'
})(LoginForm);
