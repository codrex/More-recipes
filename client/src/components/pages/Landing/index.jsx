import React from 'react';
import PropTypes from 'prop-types';
import $ from 'jquery';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { userSignup, userLogin } from '../../../actions/userActions';
import Modal from '../../common/Modal';
import LoginForm from './LoginForm';
import SignupForm from './SignupForm';
import Carousel from './Carousel';
import Cta from './CallToAction';


/**
 * Landing page with signin and signup forms
 */
class Landing extends React.Component {
/**
 * constructor
 * @param {object} props
 */
  constructor(props) {
    super(props);
    this.state = {
      signin: false,
      signup: false,
      formTitle: '',
      openModal: props.match.path !== '/',
    };
    this.signin = this.signin.bind(this);
    this.signup = this.signup.bind(this);
    this.closeModal = this.closeModal.bind(this);
    this.modalOpen = this.modalOpen.bind(this);
  }

  /**
   * componentDidMount
   * @return {undefined}
   */
  componentDidMount() {
    const { path } = this.props.match;
    if (this.state.openModal) {
      this.modalOpen(path);
    }
  }

  /**
   * componentWillReceiveProps
   * @param {object} nextProps
   * @return {undefined}
   */
  componentWillReceiveProps(nextProps) {
    const { path } = nextProps.match;

    if (path !== '/') {
      this.modalOpen(path);
    }
  }
  /**
 *
 * @param {object} nextProps
 * @return {bool} true
 * @summary redirect users to other pages after login or signup was successful
 */
  shouldComponentUpdate(nextProps) {
    const { authenticated, redirectTo } = nextProps.auth;
    if (authenticated && nextProps.auth !== this.props.auth) {
      this.props.history.push(redirectTo);
    }
    return true;
  }

  /**
   * @return {undfined} undefined
   */
  signin() {
    this.setState({
      signin: true,
      signup: false,
      formTitle: 'Log in to your account',
    });
  }

  /**
   * @return {undfined} undefined
   */
  signup() {
    this.setState({
      signin: false,
      signup: true,
      formTitle: 'Create an account',
    });
  }

  /**
   * @return {undfined} undefined
   */
  closeModal() {
    const { history } = this.props;
    this.setState({
      signin: false,
      signup: false,
      formTitle: '',
      openModal: '/'
    });
    history.push('/');
  }

  /**
  * @param {string} path
  * @return {undfined} undefined
  */
  modalOpen(path) {
    if (path === '/create-account') {
      this.signup();
    } else if (path === '/login') {
      this.signin();
    }
    const modal = $('div#modal');
    if (modal.modal) modal.modal('show');
  }

  /**
   * render
   * @return {object} object
   */
  render() {
    const { loading, history, actions } = this.props;
    return (
      <div>
        <Carousel className={this.state.formTitle && 'blur'}>
          <Cta
            signin={this.signin}
            signup={this.signup}
            push={history.push}
          />
        </Carousel>
        <Modal
          id="modal"
          center
          rightBtnText="Login"
          title={this.state.formTitle}
          closeBtnClicked={this.closeModal}
          openModal={this.state.openModal}
        >
          {
            this.state.signin &&
              <LoginForm
                login={actions.loginAction}
                loading={loading}
              />
          }
          {
            this.state.signup &&
            <SignupForm
              signup={actions.signupAction}
              loading={loading}
            />
          }
          {this.state.signin &&
            <p className="form-text">
              {"Don't have an account?"}
              <span
                role="button"
                tabIndex="0"
                onClick={() => {
                  this.signup();
                  history.push('/create-account');
                }}
              >
                <b>
                  Sign up
                </b>
              </span>
            </p>
          }
          {this.state.signup &&
            <p className="form-text">
              Already have an account?
              <span
                role="button"
                tabIndex="0"
                onClick={() => {
                  this.signin();
                  history.push('/login');
                }}
              >
                <b >
                  Sign in
                </b>
              </span>
            </p>
          }
        </Modal>
      </div>
    );
  }
}

Landing.propTypes = {
  actions: PropTypes.objectOf(PropTypes.shape).isRequired,
  history: PropTypes.objectOf(PropTypes.shape).isRequired,
  auth: PropTypes.objectOf(PropTypes.shape).isRequired,
  loading: PropTypes.bool.isRequired,
  match: PropTypes.objectOf(PropTypes.shape).isRequired
};

const mapDispatchToProps = dispatch => (
  {
    actions: {
      loginAction: bindActionCreators(userLogin, dispatch),
      signupAction: bindActionCreators(userSignup, dispatch),
    },
  }
);

const mapStateToProps = state => (
  {
    auth: state.auth,
    loading: state.networkRequest.loading,
  }

);
export { Landing as PureLanding };

export default connect(mapStateToProps, mapDispatchToProps)(Landing);
