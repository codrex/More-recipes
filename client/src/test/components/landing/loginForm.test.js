/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import toJson from 'enzyme-to-json';
import { PureLoginForm } from '../../../components/pages/Landing/LoginForm';

const props = {
  login: jest.fn(),
  handleSubmit: jest.fn(),
  submitting: false
};

describe('login form component ', () => {
  test('should render a login form', () => {
    const component = shallow(<PureLoginForm {...props} />);
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
    expect(component.find('Form').length).toBe(1);
    expect(component.find('#username').length).toBe(1);
    expect(component.find('#password').length).toBe(1);
    expect(tree).toBeInstanceOf(Object);
  });
});
