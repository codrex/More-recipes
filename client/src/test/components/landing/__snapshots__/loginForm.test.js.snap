// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login form component  expected to match snapshot when submitting props is false 1`] = `
<Form
  action=""
  className=""
  disabled={false}
  id="submit"
  lg={true}
  method=""
  onSubmit={[Function]}
  primary={false}
  primaryInverse={false}
  secondary={true}
  secondaryInverse={false}
  style={Object {}}
  submitBtnText="Login"
>
  <Field
    component={[Function]}
    id="username"
    name="username"
    placeholder="Enter username"
    type="text"
    validate={[Function]}
  />
  <Field
    component={[Function]}
    id="password"
    name="password"
    placeholder="Enter password"
    type="password"
    validate={[Function]}
  />
</Form>
`;

exports[`login form component  expected to match snapshot when submitting props is true 1`] = `
<Form
  action=""
  className=""
  disabled={true}
  id="submit"
  lg={true}
  method=""
  onSubmit={[Function]}
  primary={false}
  primaryInverse={false}
  secondary={true}
  secondaryInverse={false}
  style={Object {}}
  submitBtnText="loading..."
>
  <Field
    component={[Function]}
    id="username"
    name="username"
    placeholder="Enter username"
    type="text"
    validate={[Function]}
  />
  <Field
    component={[Function]}
    id="password"
    name="password"
    placeholder="Enter password"
    type="password"
    validate={[Function]}
  />
</Form>
`;
